### PR TITLE
[JIT] Resolve string literal type annotations using `Resolver::resolveType`

### LINKED ISSUE #47570: [jit] cannot save a model that uses "Device" type
## To Reproduce
```python
import torch
from torch import nn

class Wrapper():
    def __init__(self, t):
        self.t = t
    def to(self, device: "Device" = None):  # noqa
        return self.t.to(device=device)
class A(nn.Module):
    def forward(self):
        return Wrapper(torch.rand(4, 4))

scripted = torch.jit.script(A())
scripted.save("tmp.pt")
```
prints
```
  File "a.py", line 14, in <module>
    scripted.save("tmp.pt")
  File "/usr/lib/python3.8/site-packages/torch/jit/_script.py", line 487, in save
    return self._c.save(*args, **kwargs)
RuntimeError: Unknown device: 16. If you have recently updated the caffe2.proto file to add a new device type, did you forget to update the DeviceTypeName() function to reflect such recent changes?
Exception raised from DeviceTypeName at ../c10/core/DeviceType.cpp:33 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x72 (0x7f11a7e857c2 in /usr/lib/python3.8/site-packages/torch/lib/libc10.so)
frame #1: <unknown function> + 0x14efc (0x7f11a7e53efc in /usr/lib/python3.8/site-packages/torch/lib/libc10.so)
frame #2: c10::Device::str[abi:cxx11]() const + 0x32 (0x7f11a7e5e892 in /usr/lib/python3.8/site-packages/torch/lib/libc10.so)
frame #3: c10::operator<<(std::ostream&, c10::Device const&) + 0x22 (0x7f11a7e5eab2 in /usr/lib/python3.8/site-packages/torch/lib/libc10.so)
frame #4: c10::IValue::repr(std::ostream&, std::function<bool (std::ostream&, c10::IValue const&)>) const + 0x16f (0x7f12280bd24f in /usr/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #5: <unknown function> + 0x30a9018 (0x7f122a083018 in /usr/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #6: <unknown function> + 0x30b2ad8 (0x7f122a08cad8 in /usr/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #7: <unknown function> + 0x30b5533 (0x7f122a08f533 in /usr/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #8: <unknown function> + 0x30b5dba (0x7f122a08fdba in /usr/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #9: <unknown function> + 0x324a91c (0x7f122a22491c in /usr/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #10: <unknown function> + 0x324aca6 (0x7f122a224ca6 in /usr/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #11: <unknown function> + 0x324c6b8 (0x7f122a2266b8 in /usr/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #12: torch::jit::ExportModule(torch::jit::Module const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&, bool, bool) + 0x30c (0x7f122a221b7c in /usr/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #13: <unknown function> + 0xa61625 (0x7f122e96a625 in /usr/lib/python3.8/site-packages/torch/lib/libtorch_python.so)
frame #14: <unknown function> + 0x724eb2 (0x7f122e62deb2 in /usr/lib/python3.8/site-packages/torch/lib/libtorch_python.so)
<omitting python frames>
frame #31: __libc_start_main + 0xf2 (0x7f1260c3e152 in /usr/lib/libc.so.6)
```

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
ollecting environment information...
PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 10.2.0
Clang version: 10.0.1 
CMake version: version 3.18.4

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1070
Nvidia driver version: 455.38
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.0.4
/usr/lib/libcudnn_adv_infer.so.8.0.4
/usr/lib/libcudnn_adv_train.so.8.0.4
/usr/lib/libcudnn_cnn_infer.so.8.0.4
/usr/lib/libcudnn_cnn_train.so.8.0.4
/usr/lib/libcudnn_ops_infer.so.8.0.4
/usr/lib/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

```

cc @gmagogsfm

### PR DESCRIPTION
**Pull request by SplitInfinity**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47731 [JIT] Resolve string literal type annotations using `Resolver::resolveType`**

**Summary**
This commit modifies `ScriptTypeParser::parseTypeFromExpr` so that
string literal type annotations are resolved using
`Resolver::resolveType`. At present, they are parsed in
`parseBaseTypeName`, which inadvertently allows any key from
`string_to_type_lut` to be used as a string literal type annotation.

**Test Plan**
Existing unit tests (most notably
`TestClassType.test_self_referential_method` which tests the main
feature, self-referential class type annotations, that make use of
string literal type annotations).

**Fixes**
This commit fixes #47570.

Differential Revision: [D24934717](https://our.internmc.facebook.com/intern/diff/D24934717)

### CODE REVIEW

File: torch/csrc/jit/frontend/script_type_parser.cpp [hunk 1]
```diff
@@ -190,6 +187,15 @@ TypePtr ScriptTypeParser::parseTypeFromExprImpl(const Expr& expr) const {
     }
     return subscriptToType(*value_name, subscript);
 
+  } else if (expr.kind() == TK_STRINGLITERAL) {
+    auto type_name = StringLiteral(expr).text();
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/script_type_parser.cpp:191]: nit and optional:

Though `auto` will be deduced correctly, it is better to still include `const` and `&` for better readability.

**Review by gmagogsfm — APPROVED** [review]

**Comment by jeffdaily** [comment]
ROCm CI failure due to CI script update. Will retest.

**Comment by ppwwyyxx** [comment]
Oops I thought supporting "Device" in https://github.com/pytorch/pytorch/issues/47570 is a feature, not bug :laughing: 
Now the problem is, with this "feature" removed, and because of https://github.com/pytorch/pytorch/issues/47405, we can no longer annotate a method that uses device type in argument. So it actually causes a regression.

**Comment by SplitInfinity** [comment]
I fixed #47405 in #47734.

**Comment by ppwwyyxx** [comment]
That's great! Didn't notice this fix is also coming

**Pull request closed without merging**